leader-election-impl/src/main/java/com/palantir/leader/proxy/AwaitingLeadershipProxy.java

```diff
@@ -112,7 +116,8 @@ protected Object handleInvocation(Object proxy, Method method, Object[] args) th
                     if (leading == StillLeadingStatus.NOT_LEADING || leading == StillLeadingStatus.NO_QUORUM) {
                         return Futures.submitAsync(
                                 () -> {
-                                    leadershipStateManager.handleNotLeading(leadershipToken, null /* cause */);
+                                    leadershipStateManager.invalidateStateOnLostLeadership(
+                                            leadershipToken, null /* cause */);
                                     throw new AssertionError("should not reach here");
                                 },
```

Comment:
I'm curious why we queue this to be done async rather than invalidate immediately and return the `NotCurrentLeaderException`. We could make `invalidateStateOnLostLeadership` return `NotCurrentLeaderException` to simplify call sites, e.g. https://github.com/palantir/atlasdb/pull/6828

```
                        return Futures.immediateFailedFuture(leadershipStateManager.invalidateStateOnLostLeadership(
                                leadershipToken, null /* cause */));
```

Reply:
Hmm, I honestly don't know - this is pretty old. _I think_ it's safe to just invalidate the state in a synchronous way - I checked the code path and things look cheap - though given this was intended to deal with the PDS and we need to get a solution in soon plus I'm OOTO I plan to go ahead with this for now.

I think I'll merge this first, but we should re-aim #6828 at develop and get it in as well. @mdaudali / @LucasIME please do review part 2 while I'm out!